packages/kbn-pm/src/utils/watch.ts

```diff
@@ -0,0 +1,61 @@
+import { Observable, Subject } from 'rxjs';
+
+/**
+ * Number of milliseconds we wait before we fall back to the default watch handler.
+ */
+const defaultHandlerDelay = 3000;
+
+/**
+ * If default watch handler is used, then it's the number of milliseconds we wait for
+ * any build output before we consider watch task ready.
+ */
+const defaultHandlerReadinessTimeout = 2000;
+
+function getWatchHandlers(buildOutput$: Observable<string>) {
+  const typescriptHandler = buildOutput$
+    .first(data => data.includes('$ tsc'))
+    .map(() =>
+      buildOutput$
+        .first(data => data.includes('Compilation complete.'))
+        .mapTo('tsc')
+    );
+
+  const webpackHandler = buildOutput$
+    .first(data => data.includes('$ webpack'))
+    .map(() =>
+      buildOutput$.first(data => data.includes('Chunk Names')).mapTo('webpack')
+    );
+
+  const defaultHandler = Observable.of(undefined)
+    .delay(defaultHandlerReadinessTimeout)
+    .map(() =>
+      buildOutput$
+        .timeout(defaultHandlerDelay)
+        .catch(() => Observable.of('timeout'))
+    );
+
+  return [typescriptHandler, webpackHandler, defaultHandler];
+}
+
+export function waitUntilWatchIsReady(stream: NodeJS.EventEmitter) {
```

Comment:
For test purposes (to speed them up), what do you think about adding an extra `opts` param that looks something like:

```diff
diff --git a/packages/kbn-pm/src/utils/watch.ts b/packages/kbn-pm/src/utils/watch.ts
index 8cb822dc20..22b68b76a4 100644
--- a/packages/kbn-pm/src/utils/watch.ts
+++ b/packages/kbn-pm/src/utils/watch.ts
@@ -11,7 +11,18 @@ const defaultHandlerDelay = 3000;
  */
 const defaultHandlerReadinessTimeout = 2000;
 
-function getWatchHandlers(buildOutput$: Observable<string>) {
+interface WatchOptions {
+  handlerDelay?: number;
+  handlerReadinessTimeout?: number;
+}
+
+function getWatchHandlers(
+  buildOutput$: Observable<string>,
+  {
+    handlerDelay = defaultHandlerDelay,
+    handlerReadinessTimeout = defaultHandlerReadinessTimeout,
+  }: WatchOptions
+) {
   const typescriptHandler = buildOutput$
     .first(data => data.includes('$ tsc'))
     .map(() =>
@@ -27,17 +38,18 @@ function getWatchHandlers(buildOutput$: Observable<string>) {
     );
 
   const defaultHandler = Observable.of(undefined)
-    .delay(defaultHandlerReadinessTimeout)
+    .delay(handlerReadinessTimeout)
     .map(() =>
-      buildOutput$
-        .timeout(defaultHandlerDelay)
-        .catch(() => Observable.of('timeout'))
+      buildOutput$.timeout(handlerDelay).catch(() => Observable.of('timeout'))
     );
 
   return [typescriptHandler, webpackHandler, defaultHandler];
 }
 
-export function waitUntilWatchIsReady(stream: NodeJS.EventEmitter) {
+export function waitUntilWatchIsReady(
+  stream: NodeJS.EventEmitter,
+  opts: WatchOptions = {}
+) {
   const buildOutput$ = new Subject<string>();
   const onDataListener = (data: Buffer) =>
     buildOutput$.next(data.toString('utf-8'));
@@ -48,7 +60,7 @@ export function waitUntilWatchIsReady(stream: NodeJS.EventEmitter) {
   stream.once('error', onErrorListener);
   stream.on('data', onDataListener);
 
-  return Observable.race(getWatchHandlers(buildOutput$))
+  return Observable.race(getWatchHandlers(buildOutput$, opts))
     .mergeMap(whenReady => whenReady)
     .finally(() => {
       stream.removeListener('data', onDataListener);
```

Which means the tests could look something like:

```js
const completionHintPromise = waitUntilWatchIsReady(buildOutputStream, {
  handlerDelay: 100,
  handlerReadinessTimeout: 50
});
```

and we wouldn't need the long timeout.